### PR TITLE
[GSOC] Linux microphone support

### PR DESCRIPTION
**Pull request by DeveloppSoft**
As part of the next part of my proposal I have been implementing the linux microphone support, no framework side, everything was already in place for the Apple support.

Testing:
-  start an handler
-  simply execute a newly compiled implant
-  `mic_start -d 10` should record audio for ten seconds, don't forget to say something
-  `listen -f <path to artifact>` should be what you said

### CODE REVIEW

File: mettle/src/stdapi/audio/linux_mic.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#include <stdio.h>
+#include <string.h>
+
+#include "mic.h"
+
+FILE *arecord;
+
+struct tlv_packet *audio_mic_list(struct tlv_handler_ctx *ctx) {
+    // Popen "cat /proc/asound/pcm | grep capture"
```
Comment (mkienow-r7) [inline mettle/src/stdapi/audio/linux_mic.c]: Is this simply a note left behind from development?

**Comment by DeveloppSoft** [comment]
Yes, can be removed if you want.

**Comment by bwatters-r7** [comment]
Is this ready for final testing/merge?  @mkienow-r7?

**Comment by mkienow-r7** [comment]
@bwatters-r7 Yes, it just needs a final test and merge

**Comment by mkienow-r7** [comment]
@DeveloppSoft I'm not seeing any of the `mic_` commands with the Linux mettle I built from this PR branch. As a sanity check I've built mettle on OSX and when connected back to a handler I have the `mic_` commands available.

**Comment by DeveloppSoft** [comment]
@mkienow-r7 thought I had the same issue 😁, but actually you just need to run a mettle implant from *my branch*, so:
- clone the repo
- `make` it
- run via `./path/to/mettle -u tcp://127.0.0.1:4444 -d2`
- everything should now be ok

**Comment by mkienow-r7** [comment]
@DeveloppSoft I'm not sure exactly what was happening there, but I appreciate the sanity check. The issue I experienced seems somehow related to copying a repo from my host system to a guest VM. I was able to successfully test recording and playback of audio from Linux mettle.

I noticed an issue with the `mic_stop` command. Given `mic_start` remains in the foreground, I'm not sure how `mic_stop` was intended to be used. Testing notes in #85 and related metasploit-framework PR do not mention it.
```
Stdapi: Mic Commands
====================

    Command       Description
    -------       -----------
    listen        listen to a saved audio recording via audio player
    mic_list      list all microphone interfaces
    mic_start     start capturing an audio stream from the target mic
    mic_stop      stop capturing audio
```

*msfconsole multi/handler*
```
meterpreter > mic_start
[*] Saving to audio file: ezDERKoy.wav
[*] Streaming started...
^C[*] Streaming stopped.
[-] Error running command mic_start: Interrupt 
meterpreter > mic_stop
[*] 10.0.2.15 - Meterpreter session 1 closed.  Reason: Died

[-] Error running command mic_stop: Rex::TimeoutError Operation timed out.
```

*mettle -d3 console output*
```
[06-26-2018 13:27:53.438s] [tlv.c:570] processing method: 'audio_mic_stop' id: '38063378287877611899968554470195'
Segmentation fault
```

File: mettle/src/stdapi/audio/linux_mic.c [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    if (arecord != NULL) {
+	rc = TLV_RESULT_SUCCESS;
+    }
+
+    return tlv_packet_response_result(ctx, rc);
+}
+
+// Kill the `arecord` process
+struct tlv_packet *audio_mic_stop(struct tlv_handler_ctx *ctx) {
+    pclose(arecord);
```
Comment (mkienow-r7) [inline mettle/src/stdapi/audio/linux_mic.c]: Should we check `arecord != NULL` before attempting to close it?

**Comment by mkienow-r7** [comment]
@DeveloppSoft There is still an issue when executing the `mic_stop` command if run after `mic_start`. I can execute the `mic_stop` command as many times as I want in a fresh session prior to executing `mic_start`.

```
[06-26-2018 15:17:58.540s] [tlv.c:570] processing method: 'audio_mic_stop' id: '09787742420265249385522111515259'
stdout: Broken pipe
[06-26-2018 15:18:01.896s] [mettle.c:74] Heartbeat
[06-26-2018 15:18:06.895s] [mettle.c:74] Heartbeat
[06-26-2018 15:18:08.045s] [tlv.c:570] processing method: 'audio_mic_stop' id: '65341385624118461834872202513973'
Segmentation fault
```

**Comment by mkienow-r7** [comment]
Verified `mic_stop` no longer causes a segmentation fault in mettle.

**Comment by mkienow-r7** [comment]
# Release Notes

Adds Linux support for the microphone stdapi extension.

**Pull request merged**